allennlp/data/image_loader.py

```diff
@@ -1,55 +1,67 @@
 from os import PathLike
-from typing import Union, List, Callable, Optional, Dict, Any, Tuple
+from typing import Union, Sequence, Optional, Tuple
 
 import torch
 from torch import FloatTensor, IntTensor
 
-from allennlp.common.detectron import DetectronConfig, DetectronFlatParameters
 from allennlp.common.registrable import Registrable
 
 OnePath = Union[str, PathLike]
-ManyPaths = List[OnePath]
+ManyPaths = Sequence[OnePath]
 
 ImagesWithSize = Tuple[FloatTensor, IntTensor]
 
 
-class ImageLoader(Registrable, Callable[[Union[OnePath, ManyPaths]], FloatTensor]):
+class ImageLoader(Registrable):
     """
-    An `ImageLoader` is a callable that takes as input one or more filenames, and outputs an two tensors.
-    The first one contains the images and is of shape (batch, color, height, width).
-    The second one contains the image sizes and is of shape (batch, [height, width]).
+    An `ImageLoader` is a callable that takes as input one or more filenames, and outputs two
+    tensors.  The first one contains the images and is of shape (batch, color, height, width).  The
+    second one contains the image sizes and is of shape (batch, 2) (where the two dimensions contain
+    height and width).
     """
 
     default_implementation = "detectron"
 
-    def __call__(
-        self, filename_or_filenames: Union[OnePath, ManyPaths]
-    ) -> Tuple[FloatTensor, IntTensor]:
+    def __call__(self, filename_or_filenames: Union[OnePath, ManyPaths]) -> ImagesWithSize:
         if not isinstance(filename_or_filenames, list):
-            pixels, sizes = self([filename_or_filenames])
+            pixels, sizes = self([filename_or_filenames])  # type: ignore
             return pixels[0], sizes[0]
 
         from allennlp.common.file_utils import cached_path
+
         filenames = [cached_path(f) for f in filename_or_filenames]
         return self.load(filenames)
 
     def load(self, filenames: ManyPaths) -> ImagesWithSize:
         raise NotImplementedError()
 
 
+# These are in strings to avoid a costly / perhaps uninstalled import of detectron.  This typedef is
+# not inplace because it makes the line too long to include flake and mypy ignore statements inline.
+# TODO(mattg): Maybe better is to make some vision/ directory that doesn't get imported anywhere by
+# default, but allows for putting the detectron imports at the top of the file, because this isn't
+# the only place we have this issue.
```

Comment:
I'm pretty strongly against local imports, except in exceptional circumstances that involve heavy imports, like detectron or nltk or some optional dependency.  [PEP8 agrees](https://www.python.org/dev/peps/pep-0008/#imports), and that's the style guide that we follow.

I don't have a strong opinion on whether we should accept detectron as a required dependency of the library.  It was a pain for me to install on my mac, and it might be nice to not require users to install it.

Reply:
Oh, actually, I don't think we _can_ assume that it's always installed, because I don't think it's on pypi.  Maybe they'll change that someday?

Reply:
I also don't think we currently have a problem with circular imports.  Where are you encountering this problem?

Reply:
yes. I agree.

Reply:
Ok, I see the issue with circular imports, and it's because `dataset_readers` is inside of `data`.  The intended dependency graph goes something like `common -> data -> modules -> model -> training`.  Having the NLVR2 reader depend on a module is the cause of the circular dependency.  This would go away if we either (1) move the model and dataset reader into the models repo (which will probably happen eventually), or (2) move `dataset_readers` out of `data` and parallel to `model` in the dependency graph (which is a large breaking change that's unlikely to happen).

Reply:
If there is any time it could happen, it's for a 2.0 release.